### ISSUE TITLE
jq-1.6 **extremely** slow compared to 1.5

### ISSUE THREAD

**Issue by waldner** [stedolan/jq#1826]
Sample timing test:

```
# time for i in {1..1000}; do echo '{"foo":"bar"}' | ./jq-1.5-linux64 '.foo' > /dev/null; done

real	0m4.852s
user	0m1.352s
sys	0m0.564s
# time for i in {1..1000}; do echo '{"foo":"bar"}' | ./jq-1.6-linux64 '.foo' > /dev/null; done

real	0m46.192s
user	0m42.448s
sys	0m0.996s
```

Scripts that used to finish very fast with jq 1.5 are now timing out with 1.6. This is under Debian stretch 9.6, with the official binaries downloaded from jq site.

**Comment by nicowilliams** [comment]
This is quadratic linking of built-ins, which is now fixed in `master` thanks to a kind contribution by @muhmuhten.

**Comment by waldner** [comment]
Thanks. Is a stable release with the fix in sight anytime soon?

**Comment by nicowilliams** [comment]
I think this warrants a new release soon, but we'll probably wait for a few more things..net

**Comment by nicowilliams** [comment]
Ay, a few more things yet.

**Comment by pkoppstein** [comment]
The good news is that the "master" version is significantly better than 1.6.
The not-so-good news is that it's still slower than 1.5 and significantly slower than 1.4.
On a Mac:

```
u+s # version

 3.81 jq-1.4  
 5.85 jq-1.5  
30.6  jq-1.6 
10.7  jq-1.6-35-g80b064f 

```

**Comment by pkoppstein** [comment]
> Scripts that used to finish very fast with jq 1.5 are now timing out with 1.6.

I would venture to guess that jq is not being used optimally :-)

You might consider asking a usage question at stackoverflow.com with the jq tag.

**Comment by muhmuhten** [comment]
AFAICT the difference between jq 1.5 and jq 1.6 startup time comes down to a change in the binding strategy: 1.5 binds each builtin individually (in reverse order) only if used, while 1.6 parses a big block of builtins as a library, resolving all the symbols internally (the quadratic part).

The change seems to been made to facilitate make the `builtins` easy to generate, but in the process forces jq to link all the builtins every time.

Meanwhile, linking was always (and is still) quadratic.

**Comment by nicowilliams** [comment]
@muhmuhten I finally got a patch working that removes unbound instructions from a sub-tree of unbound instructions, but... it didn't help much, and cannot help much for reasons noted in the PR I opened for it.  A bit disappointing...

**Comment by muhmuhten** [comment]
Which opens up a new avenue for hacking around the slow startup time: if we can separate symbol resolution from parsing (just `block_join` in the parser, only do the equivalent of equivalent of `bind_block_referenced`), then we might be able to parse builtin.jq, build the array for `builtins/0`, and then do the binding, throwing out unreferenced blocks so we don't go through them repeatedly...

I'm not sure there are any easy ways to avoid quadratic linking within the current design, but that could go a long way toward bringing N back down to 1.5 levels.

**Comment by muhmuhten** [comment]
None of this is likely to affect any particular jq script that takes a long time, though; the impact is on shell scripts that call jq many (at least hundreds of) times.

If a single run of a particular jq script used to run "very fast" but is orders-of-magnitude slower now I'd look for 9a4576e7567dd38b91f28592b47eb6dafe0c4332-like accidental complexity changes to builtins or primitives.

(I was totally going to suspect _that particular_ issue looking at dates before I noticed the year changed between.)

**Comment by waldner** [comment]
> None of this is likely to affect any particular jq script that takes a long time, though; the impact is on shell scripts that call jq many (at least hundreds of) times.

Yes, that is indeed the case. They can probably be optimized (will look at that when I have time, I didn't wrote them), but the increased slowness is a thing nonetheless.

**Comment by nicowilliams** [comment]
@muhmuhten i am not at a computer right now, but i was thinking of having a compiler flag to delay all bindings of top-level defs when the module is a library...  Then we could bind all bulletins backwards, one at a time...

**Comment by nicowilliams** [comment]
Another idea is to keep `src/builtin.jq` but use two newlines to separate defs, then compile each separately as we used to and bind them separately too.

**Comment by fadado** [comment]
And what about SPLIT `builtin.jq` into one one core module, loaded always, and extra modules not loaded without an explicit `import`?  IMHO the first candidates to be removed from `builtin.jq` are the regexp related functions and the relational uppercased functions. Perhaps this not compatible enhanced modularity can be reached in the 2.0 jq...

**Comment by nicowilliams** [comment]
Closed by #1834.  Thanks @muhmuhten!

**Comment by jpmckinney** [comment]
In case anyone else comes across this issue, `brew install --HEAD jq` restores good performance.

FWIW, my query was against a JSON Lines file with `jq 'reduce (inputs | .awards[]?.items[]?.classification.id) as $id ({}; .[$id|tostring] += 1)' myfile.jsonl`

(I had written some Rust code to do a query, then realized I could do the same with jq, and was surprised that it was taking more than 10x longer. With the HEAD version it's only about 2x longer.)

**Comment by mjkno** [comment]
Is there someone who could explain why I still experience the performance problem?

$ wget https://github.com/jqlang/jq/releases/download/jq-1.5/jq-linux64 -O jq-1.5
$ wget https://github.com/jqlang/jq/releases/download/jq-1.6/jq-linux64 -O jq-1.6
$ chmod a+x jq-1.*

$ time for i in {1..1000}; do echo '{"foo":"bar"}' | ./jq-1.5 '.foo' > /dev/null; done

real    0m3.206s
user    **0m2.180s**
sys     0m0.472s

$ time for i in {1..1000}; do echo '{"foo":"bar"}' | ./jq-1.6 '.foo' > /dev/null; done

real    0m26.181s
user    **0m24.949s**
sys     0m1.303s

**Comment by wader** [comment]
@mjkno maybe try jq 1.7 or laster? i suspect this was fixed by https://github.com/jqlang/jq/pull/1834

**Comment by mjkno** [comment]
> [@mjkno](https://github.com/mjkno) maybe try jq 1.7 or laster? i suspect this was fixed by [#1834](https://github.com/jqlang/jq/pull/1834)

If it has been fixed, the fix doesn’t seem to be available in public binaries. Because of this I’m wondering whether the fix was actually implemented, even though the ticket is marked as closed.

Performace is good in in jq 1.7:
```bash
$ time for i in {1..1000}; do echo '{"foo":"bar"}' | ./jq-1.7 '.foo' > /dev/null; done
real    0m2.895s
user    0m1.578s
sys     0m0.661s
```

We rely heavily on java-jq, and unfortunately, upgrading from version 0.10.1 (which uses jq 1.5) isn’t an option because newer versions use jq 1.6, where performance is extremely poor.

**Comment by wader** [comment]
Ah now i see what you mean. Unfortunately i don't think there ever will be "patch release" of jq 1.6 to fix it. The fix is to update to jq 1.7 or later. I think your best shot is to convince or help out with updating java-jq to a later jq release, i see there is an issue about it https://github.com/arakelian/java-jq/issues/23
